File: core/src/main/kotlin/com/mirrly/tgproxy/core/NetworkProfile.kt
package com.mirrly.tgproxy.core

import org.json.JSONObject

enum class NetworkTransport {
    NONE,
    CELLULAR,
    WIFI,
    ETHERNET,
    VPN,
    OTHER;

    companion object {
        fun fromName(value: String): NetworkTransport =
            entries.firstOrNull { it.name == value.uppercase() } ?: OTHER
    }
}

enum class NetworkPowerMode {
    ACTIVE,
    SCREEN_OFF,
    POWER_SAVE
}

data class TransportSli(
    val measuredAtMs: Long = 0L,
    val smoothedRttMs: Long = -1L,
    val jitterMs: Long = 0L,
    val successRatePercent: Int = 100,
    val downstreamBps: Long = 0L,
    val upstreamBps: Long = 0L,
    val bufferbloatMs: Long = 0L
) {
    fun normalized(): TransportSli = copy(
        measuredAtMs = measuredAtMs.coerceAtLeast(0L),
        smoothedRttMs = smoothedRttMs.coerceAtLeast(-1L),
        jitterMs = jitterMs.coerceAtLeast(0L),
        successRatePercent = successRatePercent.coerceIn(0, 100),
        downstreamBps = downstreamBps.coerceAtLeast(0L),
        upstreamBps = upstreamBps.coerceAtLeast(0L),
        bufferbloatMs = bufferbloatMs.coerceAtLeast(0L)
    )

    internal fun toJson(): JSONObject = JSONObject().apply {
        put("measured_at_ms", measuredAtMs)
        put("smoothed_rtt_ms", smoothedRttMs)
        put("jitter_ms", jitterMs)
        put("success_rate_percent", successRatePercent)
        put("downstream_bps", downstreamBps)
        put("upstream_bps", upstreamBps)
        put("bufferbloat_ms", bufferbloatMs)
    }

    companion object {
        internal fun fromJson(json: JSONObject): TransportSli = TransportSli(
            measuredAtMs = json.optLong("measured_at_ms", 0L),
            smoothedRttMs = json.optLong("smoothed_rtt_ms", -1L),
            jitterMs = json.optLong("jitter_ms", 0L),
            successRatePercent = json.optInt("success_rate_percent", 100),
            downstreamBps = json.optLong("downstream_bps", 0L),
            upstreamBps = json.optLong("upstream_bps", 0L),
            bufferbloatMs = json.optLong("bufferbloat_ms", 0L)
        ).normalized()
    }
}

data class NetworkEnvironment(
    val validated: Boolean = false,
    val suspended: Boolean = false,
    val metered: Boolean = false,
    val roaming: Boolean = false,
    val congested: Boolean = false,
    val transport: NetworkTransport = NetworkTransport.NONE,
    val estimatedDownKbps: Int = 0,
    val estimatedUpKbps: Int = 0
) {
    companion object {
        fun unavailable(): NetworkEnvironment = NetworkEnvironment(suspended = true)
    }
}

/**
 * Versioned Kotlin → Rust snapshot used by every network-dependent policy.
 *
 * Partial producers update it through [withEnvironment], [withRuntime] and
 * [withTransportSli], so changing one concern cannot zero unrelated fields.
 */
data class NetworkProfile(
    val schemaVersion: Int = 1,
    val generation: Long = 1L,
    val validated: Boolean = false,
    val suspended: Boolean = true,
    val metered: Boolean = false,
    val roaming: Boolean = false,
    val congested: Boolean = false,
    val transport: NetworkTransport = NetworkTransport.NONE,
    val cellular: Boolean = false,
    val wifi: Boolean = false,
    val estimatedDownKbps: Int = 0,
    val estimatedUpKbps: Int = 0,
    val screenOn: Boolean = true,
    val powerSaveMode: Boolean = false,
    val powerMode: NetworkPowerMode = NetworkPowerMode.ACTIVE,
    val transportSli: TransportSli = TransportSli(),
    val happyEyeballsDelayMs: Long = 200L
) {
    val isMobile: Boolean
        get() = cellular || transport == NetworkTransport.CELLULAR

    fun withGeneration(value: Long): NetworkProfile = copy(generation = value).normalized()

    fun withEnvironment(environment: NetworkEnvironment): NetworkProfile = copy(
        validated = environment.validated,
        suspended = environment.suspended,
        metered = environment.metered,
        roaming = environment.roaming,
        congested = environment.congested,
        transport = environment.transport,
        cellular = environment.transport == NetworkTransport.CELLULAR,
        wifi = environment.transport == NetworkTransport.WIFI,
        estimatedDownKbps = environment.estimatedDownKbps,
        estimatedUpKbps = environment.estimatedUpKbps
    ).normalized()

    fun withRuntime(screenOn: Boolean, powerSaveMode: Boolean): NetworkProfile = copy(
        screenOn = screenOn,
        powerSaveMode = powerSaveMode,
        powerMode = when {
            powerSaveMode -> NetworkPowerMode.POWER_SAVE
            !screenOn -> NetworkPowerMode.SCREEN_OFF
            else -> NetworkPowerMode.ACTIVE
        }
    ).normalized()

    fun withTransportSli(value: TransportSli): NetworkProfile =
        copy(transportSli = value.normalized()).normalized()

    fun normalized(): NetworkProfile {
        val normalizedTransport = when {
            cellular -> NetworkTransport.CELLULAR
            wifi -> NetworkTransport.WIFI
            else -> transport
        }
        return copy(
            schemaVersion = 1,
            generation = generation.coerceAtLeast(1L),
            transport = normalizedTransport,
            cellular = normalizedTransport == NetworkTransport.CELLULAR,
            wifi = normalizedTransport == NetworkTransport.WIFI,
            estimatedDownKbps = estimatedDownKbps.coerceAtLeast(0),
            estimatedUpKbps = estimatedUpKbps.coerceAtLeast(0),
            powerMode = when {
                powerSaveMode -> NetworkPowerMode.POWER_SAVE
                !screenOn -> NetworkPowerMode.SCREEN_OFF
                else -> NetworkPowerMode.ACTIVE
            },
            transportSli = transportSli.normalized()
        )
    }

    fun toJsonString(): String = JSONObject().apply {
        val profile = normalized()
        put("schema_version", profile.schemaVersion)
        put("generation", profile.generation)
        put("validated", profile.validated)
        put("suspended", profile.suspended)
        put("metered", profile.metered)
        put("roaming", profile.roaming)
        put("congested", profile.congested)
        put("transport", profile.transport.name)
        put("cellular", profile.cellular)
        put("wifi", profile.wifi)
        put("estimated_down_kbps", profile.estimatedDownKbps)
        put("estimated_up_kbps", profile.estimatedUpKbps)
        put("screen_on", profile.screenOn)
        put("power_save_mode", profile.powerSaveMode)
        put("power_mode", profile.powerMode.name)
        put("transport_sli", profile.transportSli.toJson())
        put("happy_eyeballs_delay_ms", profile.happyEyeballsDelayMs)
    }.toString()

    companion object {
        fun fromJsonString(value: String): NetworkProfile {
            val json = JSONObject(value)
            val transport = NetworkTransport.fromName(json.optString("transport", "NONE"))
            return NetworkProfile(
                schemaVersion = json.optInt("schema_version", 1),
                generation = json.optLong("generation", 1L),
                validated = json.optBoolean("validated", false),
                suspended = json.optBoolean("suspended", true),
                metered = json.optBoolean("metered", false),
                roaming = json.optBoolean("roaming", false),
                congested = json.optBoolean("congested", false),
                transport = transport,
                cellular = json.optBoolean("cellular", transport == NetworkTransport.CELLULAR),
                wifi = json.optBoolean("wifi", transport == NetworkTransport.WIFI),
                estimatedDownKbps = json.optInt("estimated_down_kbps", 0),
                estimatedUpKbps = json.optInt("estimated_up_kbps", 0),
                screenOn = json.optBoolean("screen_on", true),
                powerSaveMode = json.optBoolean("power_save_mode", false),
                powerMode = runCatching {
                    NetworkPowerMode.valueOf(json.optString("power_mode", "ACTIVE"))
                }.getOrDefault(NetworkPowerMode.ACTIVE),
                transportSli = json.optJSONObject("transport_sli")?.let(TransportSli::fromJson)
                    ?: TransportSli(),
                happyEyeballsDelayMs = json.optLong("happy_eyeballs_delay_ms", 200L)
            ).normalized()
        }
    }
}

File: mirrlyengine/src/network_profile.rs
use crate::config::MOBILE_NETWORK;
use once_cell::sync::Lazy;
use parking_lot::RwLock;
use serde::{Deserialize, Serialize};
use std::sync::atomic::Ordering;
use std::net::IpAddr;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(default)]
pub struct TransportSli {
    pub measured_at_ms: u64,
    pub smoothed_rtt_ms: i64,
    pub jitter_ms: u64,
    pub success_rate_percent: u8,
    pub downstream_bps: u64,
    pub upstream_bps: u64,
    pub bufferbloat_ms: u64,
}

impl Default for TransportSli {
    fn default() -> Self {
        Self {
            measured_at_ms: 0,
            smoothed_rtt_ms: -1,
            jitter_ms: 0,
            success_rate_percent: 100,
            downstream_bps: 0,
            upstream_bps: 0,
            bufferbloat_ms: 0,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(default)]
pub struct NetworkProfile {
    pub schema_version: u32,
    pub generation: u64,
    pub validated: bool,
    pub suspended: bool,
    pub metered: bool,
    pub roaming: bool,
    pub congested: bool,
    pub transport: String,
    pub cellular: bool,
    pub wifi: bool,
    pub estimated_down_kbps: u32,
    pub estimated_up_kbps: u32,
    pub screen_on: bool,
    pub power_save_mode: bool,
    pub power_mode: String,
    pub transport_sli: TransportSli,
    /// Committed by the Kotlin FSM together with socket options; 0 is legacy mode.
    pub happy_eyeballs_delay_ms: u64,
    /// DUAL_STACK, IPV4_ONLY, or IPV6_FIRST. Committed by the advanced UI.
    pub ip_family_preference: String,
    /// Explicit WebSocket idle interval, in seconds. Zero is legacy adaptive mode.
    pub websocket_keep_alive_seconds: u64,
}

impl Default for NetworkProfile {
    fn default() -> Self {
        Self {
            schema_version: 1,
            generation: 1,
            validated: false,
            suspended: true,
            metered: false,
            roaming: false,
            congested: false,
            transport: "NONE".to_string(),
            cellular: false,
            wifi: false,
            estimated_down_kbps: 0,
            estimated_up_kbps: 0,
            screen_on: true,
            power_save_mode: false,
            power_mode: "ACTIVE".to_string(),
            transport_sli: TransportSli::default(),
            happy_eyeballs_delay_ms: 0,
            ip_family_preference: "DUAL_STACK".to_string(),
            websocket_keep_alive_seconds: 30,
        }
    }
}

impl NetworkProfile {
    /// Measurements are telemetry, not a new routing policy. Invalidating in-flight
    /// work for every SLI/bandwidth sample starves probes and background scans.
    fn invalidates_background_results(&self, previous: &Self) -> bool {
        self.generation != previous.generation
            || self.validated != previous.validated
            || self.suspended != previous.suspended
            || self.transport != previous.transport
            || self.metered != previous.metered
            || self.roaming != previous.roaming
            || self.power_mode != previous.power_mode
            || self.happy_eyeballs_delay_ms != previous.happy_eyeballs_delay_ms
            || self.ip_family_preference != previous.ip_family_preference
            || self.websocket_keep_alive_seconds != previous.websocket_keep_alive_seconds
    }

    fn normalized(mut self) -> Self {
        self.schema_version = 1;
        self.generation = self.generation.max(1);
        self.transport = self.transport.trim().to_ascii_uppercase();
        if self.cellular {
            self.transport = "CELLULAR".to_string();
        } else if self.wifi {
            self.transport = "WIFI".to_string();
        }
        match self.transport.as_str() {
            "CELLULAR" => {
                self.cellular = true;
                self.wifi = false;
            }
            "WIFI" => {
                self.cellular = false;
                self.wifi = true;
            }
            "NONE" | "ETHERNET" | "VPN" | "OTHER" => {
                self.cellular = false;
                self.wifi = false;
            }
            _ => {
                self.transport = "OTHER".to_string();
                self.cellular = false;
                self.wifi = false;
            }
        }
        self.transport_sli.success_rate_percent = self.transport_sli.success_rate_percent.min(100);
        self.power_mode = if self.power_save_mode {
            "POWER_SAVE".to_string()
        } else if !self.screen_on {
            "SCREEN_OFF".to_string()
        } else {
            "ACTIVE".to_string()
        };
        self.ip_family_preference = match self.ip_family_preference.trim().to_ascii_uppercase().as_str() {
            "IPV4_ONLY" => "IPV4_ONLY".to_string(),
            "IPV6_FIRST" => "IPV6_FIRST".to_string(),
            _ => "DUAL_STACK".to_string(),
        };
        self.websocket_keep_alive_seconds = self.websocket_keep_alive_seconds.clamp(15, 60);
        self
    }
}

static EFFECTIVE_NETWORK_PROFILE: Lazy<RwLock<NetworkProfile>> =
    Lazy::new(|| RwLock::new(NetworkProfile::default()));

pub fn apply_profile(profile: NetworkProfile) -> NetworkProfile {
    let effective = profile.normalized();
    crate::generation_guard::change_profile(effective.generation, || {
        let (previous_generation, changed, policy_changed) = {
            let mut guard = EFFECTIVE_NETWORK_PROFILE.write();
            let previous_generation = guard.generation;
            let changed = *guard != effective;
            let policy_changed = effective.invalidates_background_results(&guard);
            if changed {
                *guard = effective.clone();
            }
            (previous_generation, changed, policy_changed)
        };

        if changed {
            MOBILE_NETWORK.store(effective.cellular, Ordering::Relaxed);
            if effective.generation != previous_generation {
                crate::timeline::set_network_generation(effective.generation);
                crate::budget::DIAL_BUDGET.notify_generation_change(effective.generation);
                crate::balancer::BALANCER.write().notify_generation_change(effective.generation);
            }
        }
        (effective, policy_changed)
    })
    .unwrap_or_else(get_profile)
}

pub fn set_profile_json(json: &str) -> Result<NetworkProfile, String> {
    let parsed: NetworkProfile =
        serde_json::from_str(json).map_err(|error| format!("invalid network profile: {error}"))?;
    Ok(apply_profile(parsed))
}

pub fn get_profile() -> NetworkProfile {
    EFFECTIVE_NETWORK_PROFILE.read().clone()
}

pub fn current_generation() -> u64 {
    EFFECTIVE_NETWORK_PROFILE.read().generation
}

/// Filters every outbound resolver result at the engine boundary. This applies
/// equally to MTProto, SOCKS5 Worker WSS, VLESS and VPN dial paths.
pub fn filter_ip_family(ips: Vec<IpAddr>) -> Vec<IpAddr> {
    match get_profile().ip_family_preference.as_str() {
        "IPV4_ONLY" => ips.into_iter().filter(|ip| ip.is_ipv4()).collect(),
        "IPV6_FIRST" => {
            let mut v6 = Vec::new();
            let mut v4 = Vec::new();
            for ip in ips {
                if ip.is_ipv6() { v6.push(ip) } else { v4.push(ip) }
            }
            v6.extend(v4);
            v6
        }
        _ => ips,
    }
}

pub fn get_profile_json() -> String {
    serde_json::to_string(&get_profile()).unwrap_or_else(|_| "{}".to_string())
}

pub fn update_generation(generation: u64) {
    if generation == 0 {
        return;
    }
    let mut profile = get_profile();
    profile.generation = generation;
    apply_profile(profile);
}

pub fn update_legacy_transport(is_mobile: bool) {
    let mut profile = get_profile();
    profile.cellular = is_mobile;
    profile.wifi = !is_mobile;
    profile.transport = if is_mobile { "CELLULAR" } else { "WIFI" }.to_string();
    apply_profile(profile);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn profile_json_round_trip_preserves_every_dimension() {
        let profile = NetworkProfile {
            generation: 7,
            validated: true,
            suspended: false,
            metered: true,
            roaming: true,
            congested: true,
            transport: "CELLULAR".to_string(),
            cellular: true,
            estimated_down_kbps: 50_000,
            estimated_up_kbps: 12_000,
            screen_on: false,
            power_save_mode: true,
            transport_sli: TransportSli {
                measured_at_ms: 123,
                smoothed_rtt_ms: 88,
                jitter_ms: 14,
                success_rate_percent: 93,
                downstream_bps: 2_000_000,
                upstream_bps: 300_000,
                bufferbloat_ms: 31,
            },
            ..NetworkProfile::default()
        }
        .normalized();

        let json = serde_json::to_string(&profile).unwrap();
        let decoded: NetworkProfile = serde_json::from_str(&json).unwrap();
        assert_eq!(decoded, profile);
    }

    #[test]
    fn changing_power_field_preserves_transport_and_sli() {
        let original = NetworkProfile {
            generation: 9,
            validated: true,
            transport: "WIFI".to_string(),
            wifi: true,
            estimated_down_kbps: 90_000,
            transport_sli: TransportSli {
                smoothed_rtt_ms: 42,
                jitter_ms: 6,
                ..TransportSli::default()
            },
            ..NetworkProfile::default()
        }
        .normalized();
        let updated = NetworkProfile {
            screen_on: false,
            ..original.clone()
        }
        .normalized();

        assert_eq!(updated.generation, original.generation);
        assert_eq!(updated.transport, original.transport);
        assert_eq!(updated.estimated_down_kbps, original.estimated_down_kbps);
        assert_eq!(updated.transport_sli, original.transport_sli);
    }
}
